drivers/at86rf2xx/aes_spi.c

```diff
@@ -0,0 +1,356 @@
+
+#define AES_DEBUG(...)  DEBUG("[at86rf2xx_aes_spi]: "__VA_ARGS__)
+
+#define AT86RF2XX_CMD_SRAM_READ             (0b00000000)
+#define AT86RF2XX_CMD_SRAM_WRITE            (0b01000000)
+
+static inline
+void at86rf2xx_spi_get_bus(const at86rf2xx_t *dev)
+{
+    spi_acquire(dev->params.spi, dev->params.cs_pin, SPI_MODE_0, dev->params.spi_clk);
+}
+
+static inline
+void at86rf2xx_spi_release_bus(const at86rf2xx_t *dev)
+{
+    spi_release(dev->params.spi);
+}
+
+static inline
+uint8_t _aes_status(at86rf2xx_t *dev)
+{
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      AT86RF2XX_CMD_SRAM_READ);
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      AT86RF2XX_REG__AES_STATUS);
+    return spi_transfer_byte(dev->params.spi, dev->params.cs_pin, false, 0);
+}
+
+static inline
+uint8_t _aes_wait_for_result(at86rf2xx_t *dev)
+{
+    uint8_t status;
+    xtimer_usleep(AT86RF2XX_AES_DELAY_US);
+    if ((status = _aes_status(dev)) & AT86RF2XX_AES_STATUS_MASK__AES_ER) {
+        AES_DEBUG("ERROR\n");
+        return status;
+    }
+    while (!(status & AT86RF2XX_AES_STATUS_MASK__AES_DONE)) {
+        AES_DEBUG("status: %02x\n", status);
+        status = _aes_status(dev);
+    }
+    return status;
+}
+
+static inline
+void _aes_open_read(at86rf2xx_t *dev, uint8_t addr)
+{
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      AT86RF2XX_CMD_SRAM_READ);
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      addr);
+}
+
+static inline
+void _aes_open_write(at86rf2xx_t *dev, uint8_t addr)
+{
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      AT86RF2XX_CMD_SRAM_WRITE);
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, true,
+                      addr);
+}
+
+static inline
+void _aes_transfer_bytes(at86rf2xx_t *dev, bool cont, const void* out,
+                         void* in, size_t len)
+{
+    spi_transfer_bytes(dev->params.spi, dev->params.cs_pin, cont, out,
+                       in, len);
+}
+
+static inline
+void _aes_save_key(at86rf2xx_t *dev, uint8_t cfg, uint8_t key[AT86RF2XX_AES_BLOCK_SIZE])
+{
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, false, &cfg, NULL, sizeof(cfg));
+    _aes_open_read(dev, AT86RF2XX_REG__AES_KEY_START);
+    _aes_transfer_bytes(dev, false, NULL, key, AT86RF2XX_AES_KEY_LENGTH);
+}
+
+static inline
+void _aes_transfer_block(at86rf2xx_t *dev, uint8_t cfg, uint8_t mirror,
+                         const aes_block_t src, aes_block_t dst)
+{
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, true, &cfg, NULL, sizeof(cfg));
+    _aes_transfer_bytes(dev, true, src, NULL, 1);
+    _aes_transfer_bytes(dev, true, src + 1, dst, AT86RF2XX_AES_BLOCK_SIZE - 1);
+    _aes_transfer_bytes(dev, false, &mirror,
+                        dst ? dst + AT86RF2XX_AES_BLOCK_SIZE - 1 : NULL, 1);
```

Comment:
Please add a short comment as to what is happening here.

Reply:
Ok

Reply:
I said *short*, but I'm not complaining :smile: